### PR TITLE
extended param interface (now handling param clock & post ids), impro…

### PR DESCRIPTION
**Pull request by MatthiasSeeber**
…ved code readability

### CODE REVIEW

File: audio-engine/src/synth/c15-audio-engine/dsp_host.cpp [hunk 1]
```diff
@@ -101,27 +101,29 @@ void dsp_host::loadInitialPreset()
...
   /* provide indices for items, voices and parameters */
-  uint32_t i, v, p;
+  uint32_t i, v;
+  std::vector<uint32_t>* list;
   /* first: evaluate slow clock status */
   if(m_clockPosition[3] == 0)
   {
     /* render slow mono parameters and perform mono post processing */
-    for(p = 0; p < m_params.m_clockIds.m_data[3].m_data[0].m_length; p++)
+    list = &m_params.m_parameters.getClockIds(PARAM_SLOW, PARAM_MONO);
```
Comment (hhoegelo) [inline audio-engine/src/synth/c15-audio-engine/dsp_host.cpp]: Consider the more c++-ish way, using a range based loop:

for(const auto &e : m_params.m_parameters.getClockIds(PARAM_SLOW, PARAM_MONO))
{	    
    i = m_params.getHead(e).m_index;
    m_params.tickItem(i);
}

File: audio-engine/src/synth/c15-audio-engine/pe_utilities.h [hunk 2]
```diff
@@ -87,3 +70,35 @@ struct dual_env_id_list
...
+};
+
+struct new_dual_clock_id_list
+{
+    inline void add(uint32_t _spreadType, uint32_t _clockType, uint32_t _polyType, uint32_t _id)
+    {
+        m_data[_spreadType][_clockType][_polyType].push_back(_id);
+    }
+
+    inline std::vector<uint32_t> &get(uint32_t _spreadType, uint32_t _clockType, uint32_t _polyType)
```
Comment (hhoegelo) [inline audio-engine/src/synth/c15-audio-engine/pe_utilities.h]: consider const access:

inline const std::vector<uint32_t> &get(uint32_t _spreadType, uint32_t _clockType, uint32_t _polyType) const

File: audio-engine/src/synth/c15-audio-engine/pe_utilities.h [hunk 3]
```diff
@@ -87,3 +70,35 @@ struct dual_env_id_list
...
 };
+
+struct new_clock_id_list
+{
+    inline void add(uint32_t _clockType, uint32_t _polyType, uint32_t _id)
+    {
+        m_data[_clockType][_polyType].push_back(_id);
+    }
+
+    inline std::vector<uint32_t> &get(uint32_t _clockType, uint32_t _polyType)
```
Comment (hhoegelo) [inline audio-engine/src/synth/c15-audio-engine/pe_utilities.h]: consider const access:

inline const std::vector<uint32_t> &get(uint32_t _clockType, uint32_t _polyType) const

**Comment by MatthiasSeeber** [comment]
now implemented:
- range based loops (dsp_host::init, dsp_host::tickMain, param_engine::postProcess[..])
- const access (id_list::get, param_engine::getClockIds, param_engine::getPostIds)
- loops: minimal variables scope (i, v, p) when possible (dsp_host::init, dsp_host::tickMain, param_engine::postProcess[..])

todo:
- back to fixed size arrays for id lists (hopefully providing non-fragmented memory and faster access)
- direct iteration of param_body rendering (dsp_host::tickMain)

File: audio-engine/src/synth/c15-audio-engine/dsp_host.cpp [hunk 4]
```diff
@@ -132,45 +129,44 @@ void dsp_host::tickMain()
...
+        auto i = m_params.getHead(it).m_index;
+        m_params.tickItem(i);
     }
     m_params.postProcessMono_fast(m_parameters.bindToVoice(0));
     /* render fast poly parameters and perform poly fast post processing */
-    for(v = 0; v < m_voices; v++)
+    for(uint32_t v = 0; v < m_voices; v++)
     {
-      for(p = 0; p < m_params.m_clockIds.m_data[2].m_data[1].m_length; p++)
+      for(auto &it : m_params.m_parameters.getClockIds(PARAM_FAST, PARAM_POLY))
```
Comment (hhoegelo) [inline audio-engine/src/synth/c15-audio-engine/dsp_host.cpp]: This does not look right to me: for each voice, you process all (fast, poly)-params, but the inner loop seems not to depend on the voice at all. As far as I can see, you could as well process the params once, then loop over the voices  for postProcessPoly_fast only.

Comment (MatthiasSeeber) [reply audio-engine/src/synth/c15-audio-engine/dsp_host.cpp]: Oh, yes - i forgot to add _voiceId to the m_index in the poly contexts.

**Pull request merged**